external-crates/move/crates/move-compiler/src/hlir/detect_dead_code.rs

```diff
@@ -0,0 +1,770 @@
+        //  statements with effects
+        // -----------------------------------------------------------------------------------------
+        E::Assign(_, _, rhs) => {
+            if let Some(rhs_control_flow) = value(context, rhs) {
+                context.report_value_error(rhs_control_flow);
+            }
+            None
+        }
+        E::Mutate(lhs, rhs) => {
+            if let Some(rhs_control_flow) = value(context, rhs) {
+                context.report_value_error(rhs_control_flow);
+            } else if let Some(lhs_control_flow) = value(context, lhs) {
+                context.report_value_error(lhs_control_flow);
+            }
+            None
+        }
+
+        // -----------------------------------------------------------------------------------------
+        // valued expressions -- when these occur in statement position need their children
+        // unravelled to find any embedded, divergent operations.
+        // -----------------------------------------------------------------------------------------
+        E::ModuleCall(_)
+        | E::Builtin(_, _)
+        | E::Vector(_, _, _, _)
+        | E::Dereference(_)
+        | E::UnaryExp(_, _)
+        | E::BinopExp(_, _, _, _)
+        | E::Pack(_, _, _, _)
+        | E::ExpList(_)
+        | E::Borrow(_, _, _)
+        | E::TempBorrow(_, _)
+        | E::Cast(_, _)
+        | E::Annotate(_, _)
+        | E::BorrowLocal(_, _)
+        | E::Constant(_, _)
+        | E::Move { .. }
+        | E::Copy { .. }
+        | E::Spec(..)
+        | E::UnresolvedError => value(context, e),
+
+        E::Value(_) | E::Unit { .. } => None,
+
+        // -----------------------------------------------------------------------------------------
+        // odds and ends -- things we need to deal with but that don't do much
+        // -----------------------------------------------------------------------------------------
+        E::Use(_) => panic!("ICE unexpanded use"),
+    }
+}
+
+fn statement_block(
+    context: &mut Context,
+    seq: &T::Sequence,
+    stmt_pos: bool,
+    skip_last: bool,
+) -> Option<ControlFlow> {
+    use T::SequenceItem_ as S;
+
+    // if we're in statement position, we need to check for a trailing semicolon error
+    // this code does that by noting a trialing unit and then proceeding as if we are not in
+    // statement position.
+    if stmt_pos && has_trailing_unit(seq) {
+        let last = seq.iter().last();
+        let result = statement_block(context, seq, false, true);
+        return if let (Some(control_flow), Some(sp!(_, S::Seq(entry)))) = (result, last) {
+            context.report_statement_tail_error(control_flow, entry);
+            None
+        } else {
+            None
+        };
+    }
+
+    let iterator = if skip_last {
+        seq.iter().skip_last().enumerate().collect::<Vec<_>>()
+    } else {
+        seq.iter().enumerate().collect::<Vec<_>>()
+    };
+    let last_ndx = iterator.iter().skip(1).len();
```

Comment:
Your iterator is actually a `Vec` so you could just do `iterator.len() - 1` here, right?

Reply:
Also probably worth not calling it iterator -- that's a bit confusing.

Reply:
The alternative is `usize::try_from(u32::try_from(iterator.len()).unwrap() - 1).unwrap()` -- unless there's an easier way to do that?

Reply:
It ends up with `usize::saturating_sub`, but yes it'll work.